File: leetcode/0912-sort-an-array/solution.cpp
#include <vector>
#include <cstdlib>
using namespace std;

// Merge sort: guaranteed O(n log n) worst case.
class Solution {
public:
    vector<int> sortArray(vector<int>& nums) {
        vector<int> a = nums;
        mergeSort(a, 0, (int)a.size() - 1);
        return a;
    }

private:
    void mergeSort(vector<int>& a, int lo, int hi) {
        if (lo >= hi) return;
        int mid = lo + (hi - lo) / 2;
        mergeSort(a, lo, mid);
        mergeSort(a, mid + 1, hi);
        merge(a, lo, mid, hi);
    }

    void merge(vector<int>& a, int lo, int mid, int hi) {
        vector<int> temp(hi - lo + 1);
        int i = lo, j = mid + 1, k = 0;
        while (i <= mid && j <= hi) {
            temp[k++] = (a[i] <= a[j]) ? a[i++] : a[j++];
        }
        while (i <= mid) temp[k++] = a[i++];
        while (j <= hi) temp[k++] = a[j++];
        for (int t = 0; t < (int)temp.size(); t++) {
            a[lo + t] = temp[t];
        }
    }
};

// Alternative: quicksort with a randomized pivot, expected O(n log n).
class QuickSortSolution {
public:
    vector<int> sortArray(vector<int>& nums) {
        vector<int> a = nums;
        quicksort(a, 0, (int)a.size() - 1);
        return a;
    }

private:
    void quicksort(vector<int>& a, int lo, int hi) {
        if (lo >= hi) return;
        int p = partition(a, lo, hi);
        quicksort(a, lo, p - 1);
        quicksort(a, p + 1, hi);
    }

    int partition(vector<int>& a, int lo, int hi) {
        int pivotIdx = lo + rand() % (hi - lo + 1);
        swap(a[pivotIdx], a[hi]);
        int pivot = a[hi];

        int i = lo;
        for (int j = lo; j < hi; j++) {
            if (a[j] < pivot) {
                swap(a[i], a[j]);
                i++;
            }
        }
        swap(a[i], a[hi]);
        return i;
    }
};

File: leetcode/0912-sort-an-array/solution.py
import random
from typing import List


class Solution:
    """Merge sort: guaranteed O(n log n) worst case."""

    def sortArray(self, nums: List[int]) -> List[int]:
        if len(nums) <= 1:
            return nums

        mid = len(nums) // 2
        left = self.sortArray(nums[:mid])
        right = self.sortArray(nums[mid:])
        return self._merge(left, right)

    def _merge(self, left: List[int], right: List[int]) -> List[int]:
        merged = []
        i = j = 0
        while i < len(left) and j < len(right):
            if left[i] <= right[j]:
                merged.append(left[i])
                i += 1
            else:
                merged.append(right[j])
                j += 1
        merged.extend(left[i:])
        merged.extend(right[j:])
        return merged


class QuickSortSolution:
    """Quicksort with a randomized pivot: expected O(n log n)."""

    def sortArray(self, nums: List[int]) -> List[int]:
        nums = nums[:]
        self._quicksort(nums, 0, len(nums) - 1)
        return nums

    def _quicksort(self, nums: List[int], lo: int, hi: int) -> None:
        if lo >= hi:
            return
        p = self._partition(nums, lo, hi)
        self._quicksort(nums, lo, p - 1)
        self._quicksort(nums, p + 1, hi)

    def _partition(self, nums: List[int], lo: int, hi: int) -> int:
        pivot_idx = random.randint(lo, hi)
        nums[pivot_idx], nums[hi] = nums[hi], nums[pivot_idx]
        pivot = nums[hi]

        i = lo
        for j in range(lo, hi):
            if nums[j] < pivot:
                nums[i], nums[j] = nums[j], nums[i]
                i += 1
        nums[i], nums[hi] = nums[hi], nums[i]
        return i
